core/src/main/java/org/elasticsearch/client/transport/TransportClientNodesService.java

```diff
@@ -535,8 +531,13 @@ public void handleException(TransportException e) {
                                         }
                                         finally {
                                             latch.countDown();
+                                            closeConnection();
                                         }
                                     }
+
+                                    void closeConnection() {
+                                        IOUtils.closeWhileHandlingException(connectionToClose);
+                                    }
```

Comment:
testing will be tricky but doable. I have some ideas here similar to what I did on `RemoteClusterConnectionTests` where we basically mock the calls to clusterstate and return a pre-build state but we can also put some sleeps into it.